File: src/main/ruby/g0201_0300/s0207_course_schedule/solution.rb
# #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Depth_First_Search
# #Breadth_First_Search #Graph #Topological_Sort #Big_O_Time_O(N)_Space_O(N)
# #2023_11_25_Time_60_ms_(95.42%)_Space_212.4_MB_(26.72%)

WHITE = 0
GRAY = 1
BLACK = 2

# @param {Integer} num_courses
# @param {Integer[][]} prerequisites
# @return {Boolean}
def can_finish(num_courses, prerequisites)
  adj = Array.new(num_courses) {[]}

  prerequisites.each do |pre|
    adj[pre[1]] << pre[0]
  end

  colors = Array.new(num_courses, WHITE)

  (0...num_courses).each do |i|
    if colors[i] == WHITE && !adj[i].empty? && has_cycle(adj, i, colors)
      return false
    end
  end

  true
end

private

def has_cycle(adj, node, colors)
  colors[node] = GRAY

  adj[node].each do |nei|
    return true if colors[nei] == GRAY
    return true if colors[nei] == WHITE && has_cycle(adj, nei, colors)
  end

  colors[node] = BLACK
  false
end

File: src/main/php/g0201_0300/s0207_course_schedule/Solution.php
<?php

namespace leetcode\g0201_0300\s0207_course_schedule;

// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Depth_First_Search
// #Breadth_First_Search #Graph #Topological_Sort #Top_Interview_150_Graph_General
// #Big_O_Time_O(N)_Space_O(N) #2023_12_21_Time_25_ms_(89.29%)_Space_22.3_MB_(100.00%)

class Solution {
    const WHITE = 0;
    const GRAY = 1;
    const BLACK = 2;

    /**
     * @param Integer $numCourses
     * @param Integer[][] $prerequisites
     * @return Boolean
     */
    public function canFinish($numCourses, $prerequisites) {
        $adj = array_fill(0, $numCourses, []);
        foreach ($prerequisites as $pre) {
            array_push($adj[$pre[1]], $pre[0]);
        }
        $colors = array_fill(0, $numCourses, self::WHITE);
        for ($i = 0; $i < $numCourses; $i++) {
            if ($colors[$i] == self::WHITE && !empty($adj[$i]) && $this->hasCycle($adj, $i, $colors)) {
                return false;
            }
        }
        return true;
    }

    private function hasCycle($adj, $node, &$colors) {
        $colors[$node] = self::GRAY;
        foreach ($adj[$node] as $nei) {
            if ($colors[$nei] == self::GRAY) {
                return true;
            }
            if ($colors[$nei] == self::WHITE && $this->hasCycle($adj, $nei, $colors)) {
                return true;
            }
        }
        $colors[$node] = self::BLACK;
        return false;
    }
}
